Sources/SourceKitLSP/Swift/SwiftLanguageService.swift

```diff
@@ -922,29 +939,19 @@ extension SwiftLanguageService {
   }
 
   public func executeCommand(_ req: ExecuteCommandRequest) async throws -> LSPAny? {
-    // TODO: If there's support for several types of commands, we might need to structure this similarly to the code actions request.
-    guard let sourceKitLSPServer else {
-      // `SourceKitLSPServer` has been destructed. We are tearing down the language
-      // server. Nothing left to do.
-      throw ResponseError.unknown("Connection to the editor closed")
-    }
-    guard let swiftCommand = req.swiftCommand(ofType: SemanticRefactorCommand.self) else {
-      throw ResponseError.unknown("semantic refactoring: unknown command \(req.command)")
-    }
-    let refactor = try await semanticRefactoring(swiftCommand)
-    let edit = refactor.edit
-    let req = ApplyEditRequest(label: refactor.title, edit: edit)
-    let response = try await sourceKitLSPServer.sendRequestToClient(req)
-    if !response.applied {
-      let reason: String
-      if let failureReason = response.failureReason {
-        reason = " reason: \(failureReason)"
-      } else {
-        reason = ""
+
+    let commandsAndHandlers: [(command: any SwiftCommand.Type, handler: SwiftCommandHandler)] = [
+      (command: SemanticRefactorCommand.self, handler: semanticRefactoring),
+      (command: ExpandMacroCommand.self, handler: expandMacro),
+    ]
+
+    for (command, handler) in commandsAndHandlers {
+      if let swiftCommand = req.swiftCommand(ofType: command) {
+        return try await handler(swiftCommand)
       }
-      logger.error("client refused to apply edit for \(refactor.title, privacy: .public)!\(reason)")
     }
-    return edit.encodeToLSPAny()
+
+    throw ResponseError.unknown("semantic refactoring: unknown command \(req.command)")
```

Comment:
`ExecuteCommand` no longer is just about semantic refactorings, right? I’d just drop the `semantic refactoring: ` prefix from the log message.